crml/cennzx-spot/src/lib.rs

```diff
@@ -55,8 +56,83 @@ pub trait Trait: frame_system::Trait + pallet_generic_asset::Trait {
 	type UnsignedIntToBalance: From<LowPrecisionUnsigned> + Into<<Self as pallet_generic_asset::Trait>::Balance>;
 }
 
+decl_error! {
+	/// Error for the generic-asset module.
+	pub enum Error for Module<T: Trait> {
+		/// Exchange pool is empty.
+		EmptyExchangePool,
+		// Insufficient trade asset reserve in exchange
+		InsufficientTradeAssetReserve,
+		// Insufficient core asset reserve in exchange
+		InsufficientCoreAssetReserve,
+		// Insufficient asset balance in buyer account
+		InsufficientBuyerTradeAssetBalance,
+		// Insufficient core asset balance in buyer account
+		InsufficientBuyerCoreAssetBalance,
+		// Insufficient asset balance in seller account
+		InsufficientSellerTradeAssetBalance,
+		// Insufficient core asset balance in seller account
+		InsufficientSellerCoreAssetBalance,
+		// Buy amount must be a positive value
+		BuyAmountNotPositive,
+		// The sale value of input is less than the required minimum.
+		SaleValueBelowRequiredMinimum,
+		// Asset sale value should be greater than zero
+		AssetSaleValueNotAboveZero,
+		// Asset to core sale price should be greater than zero
+		AssetToCorePriceNotAboveZero,
+		// Insufficient core asset balance in exchange account
+		InsufficientCoreAssetInExchangeBalance,
+		// Asset to core sale price exceeds the specified max. limit
+		AssetToCorePriceAboveMaxLimit,
+		// Core to asset sale price should be greater than zero
+		CoreToAssetPriceNotAboveZero,
+		// Core to asset sale price exceeds the specified max. limit
+		CoreToAssetPriceAboveMaxLimit,
+		// Asset to asset sale price exceeds the specified max. limit
+		AssetToAssetPriceAboveMaxLimit,
+		// Tried to overdraw liquidity
+		LiquidityTooLow,
+		// Minimum trade asset is required
+		MinimumTradeAssetIsRequired,
+		// Minimum core asset is required
+		MinimumCoreAssetIsRequired,
+		// Assets withdrawn to be greater than zero
+		AssetToWithdrawNotAboveZero,
+		// Amount of exchange asset to burn should exist
+		LiquidityToWithdrawNotAboveZero,
+		// Liquidity should exist
+		NoLiquidityToRemove,
+		// trade asset amount must be greater than zero
+		TradeAssetToAddLiquidityNotAboveZero,
```

Comment:
Seems like we can merge a few of such errors in one enum entry as "AssetNotAboveZero". I know we want to generate a different error message, we should also take the fact that the type of the error is the same into account.